### PR TITLE
Fix for issue #3802 (mail getting triggered multiple times while post…

### PR DESCRIPTION
**Pull request by danieljames-dj**
…ing results)

### CODE REVIEW

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 1]
```diff
@@ -266,6 +266,9 @@ def post_results
       unless comp.results
         return render html: "<div class='container'><div class='alert alert-warning'>No results</div></div>".html_safe
       end
+      unless comp.results_posted.nil?
```
Comment (jonatanklosko) [inline WcaOnRails/app/controllers/competitions_controller.rb]: There is no `Competition#results_posted` method. Competition has `results_posted_at` timestamp and a convenient `results_posted?` method: https://github.com/thewca/worldcubeassociation.org/blob/c89a14ae08fb8ad4815968f6c1929021d2cdda84/WcaOnRails/app/models/competition.rb#L873-L875
So the condition should be as straightforward as `if comp.results_posted?` =)

**Comment by danieljames-dj** [comment]
Hi Jonatan, thanks for your comment. It's my bad, I was about to take results_posted_at from the suggestion, but took the wrong one.
But there's one more issue: if the post is deleted, the results_posted_at won't become nil. So, if WRT wants to post the result again, it won't be possible. Have to think of a workaround for this as well. I'll fix both these issues and give another PR.
My suggestion for reposting will be - when 'post results' are clicked, first set results_posted_at to nil and then go to the URL.

**Comment by danieljames-dj** [comment]
Small change in the suggestion: if we click 'post results', if the results_posted_at is nil, we can directly go to the URL. else, we will put an alert like "Results are already posted, are you sure you want to continue? If yes, please make sure you delete the old post if not done yet." and then if WRT clicks yes, we will set results_posted_at to nil and then open the URL.

**Comment by jonatanklosko** [comment]
Sounds reasonable to me, as long as the WRT also likes it =) Not sure about setting `results_posted_at` to `nil` though, the way I see it is either behave normally when it's `nil` or show a confirmation alert otherwise (so confirming the action will just re-post the announcement and update `results_posted_at`).

**Comment by danieljames-dj** [comment]
Hi Jonatan, being new here, I'm curious to know what will be the difficulty in PHP part. Is the communication with DB difficult there?
If we can't get the confirmation there, can we get the confirmation at 'def post_results' itself? If results_posted_at is not nil, we can get a confirmation from the WRT member 'Are you sure you want to post the results again?' and if we get yes, then proceed else return?

**Comment by jonatanklosko** [comment]
The thing is we don't usually change the PHP part as it's harder to set up locally and things are just easier on the Rails side in our case. What I initially thought was showing the standard javascript confirmation "are you sure?" (based on the `results_posted_at` we could either render the button normally or with the confirmation "enabled").
I'm not sure what you mean by confirming the action in `def post_results`, we need to render a page from there, no way to just get confirmation. We could render a confirmation page with a button posting to the same URL, but with a special URL query param indicating implicit confirmation.

**Comment by danieljames-dj** [comment]
Hi, I've made the changes and added the commits.
I've followed what Sébastien suggested in a mail thread: if results_posted_at is not nil, we will just create the post, else along with creating post, we will be sending mail & updating results_posted_at.

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 2]
```diff
@@ -342,9 +342,11 @@ def post_results
           body += "#{record_strs.join(", ")}.  \n" # Trailing spaces for markdown give us a <br>
         end
       end
-      comp.update!(results_posted_at: Time.now)
-      comp.competitor_users.each { |user| user.notify_of_results_posted(comp) }
-      comp.registrations.accepted.each { |registration| registration.user.notify_of_id_claim_possibility(comp) }
+      unless comp.results_posted
```
Comment (jonatanklosko) [inline WcaOnRails/app/controllers/competitions_controller.rb]: Nit: it should be `comp.results_posted?`.

**Comment by danieljames-dj** [comment]
Hi Jonatan,

I've submitted the new PR with just one commit - https://github.com/thewca/worldcubeassociation.org/pull/4040

Sorry that I didn't understand your statement "One comment that should help in making Travis happy", may I know the meaning of same? Is it just adding a comment why the unless statement is?

**Comment by jonatanklosko** [comment]
@danieljames-dj sorry for not being clear, I just meant that the code typo I commented on was the reason why Travis build failed.

**Comment by jonatanklosko** [comment]
Superseded by #3802.

**Pull request closed without merging**